stdlib/ballerina-builtin/src/main/java/org/ballerinalang/nativeimpl/reflect/DeepEquals.java

```diff
@@ -0,0 +1,278 @@
+                        return false;
+                    }
+                }
+                break;
+            case TypeTags.ARRAY_TAG:
+            case TypeTags.ANY_TAG:
+                // TODO: This block should ideally be in the ARRAY_TAG case only. Not ANY_TAG. #4505.
+                if (lhsValue instanceof BNewArray) {
+                    BNewArray lhsArray = (BNewArray) lhsValue;
+                    BNewArray rhsArray = (BNewArray) rhsValue;
+                    if (lhsArray.size() != rhsArray.size()) {
+                        return false;
+                    }
+                    
+                    for (int i = 0; i < lhsArray.size(); i++) {
+                        if (!isDeepEqual(lhsArray.getBValue(i), rhsArray.getBValue(i))) {
+                            return false;
+                        }
+                    }
+                }
+                break;
+            case TypeTags.XML_TAG:
+                // TODO: https://www.pixelstech.net/article/1453272563-Canonicalize-XML-in-Java
+                break;
+            default:
+                return false;
+        }
+        
+        return true;
+    }
+    
+    /**
+     * Deep equals function for JSON type values.
+     *
+     * @param lhsJson The left side value.
+     * @param rhsJson The right side value.
+     * @return True if values are equal, else false.
+     */
+    private boolean isDeepEqual(JsonNode lhsJson, JsonNode rhsJson) {
+        if (lhsJson.getType() == JsonNode.Type.OBJECT) {
+            // Converting iterators to maps as iterators are ordered.
+            Iterator<Map.Entry<String, JsonNode>> lhJsonFieldsIterator = lhsJson.fields();
+            HashMap<String, JsonNode> lhJsonFields = new HashMap<>();
+            while (lhJsonFieldsIterator.hasNext()) {
+                Map.Entry<String, JsonNode> field = lhJsonFieldsIterator.next();
+                lhJsonFields.put(field.getKey(), field.getValue());
+            }
+            
+            Iterator<Map.Entry<String, JsonNode>> rhJsonFieldsIterator = rhsJson.fields();
+            HashMap<String, JsonNode> rhJsonFields = new HashMap<>();
+            while (rhJsonFieldsIterator.hasNext()) {
+                Map.Entry<String, JsonNode> field = rhJsonFieldsIterator.next();
+                rhJsonFields.put(field.getKey(), field.getValue());
+            }
+            
+            // Size of the fields should be same.
+            if (lhJsonFields.size() != rhJsonFields.size()) {
+                return false;
+            }
+            
+            if (lhJsonFields.size() > 0) {
+                for (Map.Entry<String, JsonNode> entry : lhJsonFields.entrySet()) {
+                    if (!rhJsonFields.keySet().contains(entry.getKey())) {
+                        return false;
+                    }
+                    if (!isDeepEqual(entry.getValue(), rhJsonFields.get(entry.getKey()))) {
+                        return false;
+                    }
```

Comment:
Can't we simplify the two if clauses to be one, since both are returning false for either of the conditions?

Reply:
Fixed with https://github.com/ballerina-lang/ballerina/pull/4680/commits/d202e03abb90cf8c72e87f4c5ead361106f4f4df